### PR TITLE
8265128: [REDO] Optimize Vector API slice and unslice operations

### PR DESCRIPTION
**Pull request by sviswa7**
All the slice and unslice variants that take more than one argument can benefit from already intrinsic methods on similar lines as slice(origin) and unslice(origin).

Changes include:
 * Rewrite Vector API slice/unslice using already intrinsic methods
 * Fix in library_call.cpp:inline_preconditions_checkIndex() to not modify control if intrinsification fails
 * Vector API conversion tests thresholds adjustment
 
Base Performance:
Benchmark (size) Mode Cnt Score Error Units
TestSlice.vectorSliceOrigin 1024 thrpt 5 11763.372 ± 254.580 ops/ms
TestSlice.vectorSliceOriginVector 1024 thrpt 5 599.286 ± 326.770 ops/ms
TestSlice.vectorSliceUnsliceOrigin 1024 thrpt 5 6627.601 ± 22.060 ops/ms
TestSlice.vectorSliceUnsliceOriginVector 1024 thrpt 5 401.858 ± 220.340 ops/ms
TestSlice.vectorSliceUnsliceOriginVectorPart 1024 thrpt 5 421.993 ± 231.703 ops/ms

Performance with patch:
Benchmark (size) Mode Cnt Score Error Units
TestSlice.vectorSliceOrigin 1024 thrpt 5 11792.091 ± 37.296 ops/ms
TestSlice.vectorSliceOriginVector 1024 thrpt 5 8388.174 ± 115.886 ops/ms
TestSlice.vectorSliceUnsliceOrigin 1024 thrpt 5 6662.159 ± 8.203 ops/ms
TestSlice.vectorSliceUnsliceOriginVector 1024 thrpt 5 5206.300 ± 43.637 ops/ms
TestSlice.vectorSliceUnsliceOriginVectorPart 1024 thrpt 5 5194.278 ± 13.376 ops/ms

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265128](https://bugs.openjdk.java.net/browse/JDK-8265128): [REDO] Optimize Vector API slice and unslice operations

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3804/head:pull/3804` \
`$ git checkout pull/3804`

Update a local copy of the PR: \
`$ git checkout pull/3804` \
`$ git pull https://git.openjdk.java.net/jdk pull/3804/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3804`

View PR using the GUI difftool: \
`$ git pr show -t 3804`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3804.diff">https://git.openjdk.java.net/jdk/pull/3804.diff</a>

</details>

### CODE REVIEW

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 1]
```diff
@@ -2279,6 +2276,17 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
     public abstract
     $abstractvectortype$ slice(int origin);
 
+    /*package-private*/
+    final
+    @ForceInline
+    $abstractvectortype$ sliceTemplate(int origin) {
+        Objects.checkIndex(origin, length());
+        VectorShuffle<$Boxtype$> Iota = iotaShuffle();
```
Comment (PaulSandoz) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2284]:
```suggestion
        VectorShuffle<$Boxtype$> iota = iotaShuffle();
```
use lower case first letter for variable names.

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 2]
```diff
@@ -2279,6 +2276,17 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
...
+        VectorMask<$Boxtype$> BlendMask = Iota.toVector().compare(VectorOperators.LT, (broadcast(($type$)(length() - origin))));
+        Iota = iotaShuffle(origin, 1, true);
+        return vspecies().zero().blend(this.rearrange(Iota), BlendMask);
```
Comment (PaulSandoz) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2287]: Observation: this is like `this.rearrange(iota, blendMask)` but we avoid the checking for exceptional shuffle indexes, since we know there are no such exceptional indexes.

And, of course,this method is like `slice(origin vspecies().zero())`. I wonder if this template should defer to the vector accepting tempting? Or the method on vector itself?

File: test/jdk/jdk/incubator/vector/AbstractVectorConversionTest.java [hunk 3]
```diff
@@ -48,7 +48,7 @@ public void getRunTime(ITestResult tr) {
         System.out.println(tr.getName() + " took " + time + " ms");
     }
 
-    static final int INVOC_COUNT = Integer.getInteger("jdk.incubator.vector.test.loop-iterations", 1000);
+    static final int INVOC_COUNT = Integer.getInteger("jdk.incubator.vector.test.loop-iterations", 100);
```
Comment (PaulSandoz) [inline test/jdk/jdk/incubator/vector/AbstractVectorConversionTest.java:51]: Why did you change this?

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 4]
```diff
@@ -2279,6 +2276,17 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
     public abstract
     $abstractvectortype$ slice(int origin);
 
+    /*package-private*/
+    final
+    @ForceInline
+    $abstractvectortype$ sliceTemplate(int origin) {
+        Objects.checkIndex(origin, length());
```
[the remark this answers is not in the document]
Comment (sviswa7) [reply src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2283]: Thanks for pointing this out. I will update this to Objects.checkIndex(origin, length()+1);

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template (hunk 1 above)
Comment (sviswa7) [reply src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2284]: I will make this change.

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template (hunk 2 above)
Comment (sviswa7) [reply src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2287]: Trying to optimize as much as possible:
  There is overhead in checking for exception shuffle indices. 
  The slice(origin, vector) version has one extra rearrange call.

File: test/jdk/jdk/incubator/vector/AbstractVectorConversionTest.java (hunk 3 above)
Comment (sviswa7) [reply test/jdk/jdk/incubator/vector/AbstractVectorConversionTest.java:51]: Following reasons:
The INVOC_COUNT is set as 100 for all other tests.
The conversion tests take longer time to complete than other tests.
For the conversion tests it looks like the intrinsics don't trigger sometimes and tests take longer to execute and timeout. This was one of the reasons for backout last time.

**Comment by sviswa7** [comment]
@PaulSandoz I have implemented your review comments. Please take a look.

File: test/jdk/jdk/incubator/vector/AbstractVectorConversionTest.java (hunk 3 above)
Comment (PaulSandoz) [reply test/jdk/jdk/incubator/vector/AbstractVectorConversionTest.java:51]: ok

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 5]
```diff
@@ -2250,14 +2250,11 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
...
-        int vlen = res.length;
-        int firstPart = vlen - origin;
-        System.arraycopy(a0, origin, res, 0, firstPart);
-        System.arraycopy(a1, 0, res, firstPart, origin);
-        return vectorFactory(res);
+        Objects.checkIndex(origin, length() + 1);
+        VectorShuffle<$Boxtype$> iota = iotaShuffle();
+        VectorMask<$Boxtype$> BlendMask = iota.toVector().compare(VectorOperators.LT, (broadcast(($type$)(length() - origin))));
+        iota = iotaShuffle(origin, 1, true);
+        return (($abstractvectortype$)v1).rearrange(iota).blend(this.rearrange(iota), BlendMask);
```
Comment (PaulSandoz) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2257]:
```suggestion
        return that.rearrange(iota).blend(this.rearrange(iota), blendMask);
```

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 6]
```diff
@@ -2250,14 +2250,11 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
...
-        $type$[] a1 = that.vec();
-        $type$[] res = new $type$[a0.length];
-        int vlen = res.length;
-        int firstPart = vlen - origin;
-        System.arraycopy(a0, origin, res, 0, firstPart);
-        System.arraycopy(a1, 0, res, firstPart, origin);
-        return vectorFactory(res);
+        Objects.checkIndex(origin, length() + 1);
+        VectorShuffle<$Boxtype$> iota = iotaShuffle();
+        VectorMask<$Boxtype$> BlendMask = iota.toVector().compare(VectorOperators.LT, (broadcast(($type$)(length() - origin))));
```
Comment (PaulSandoz) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2255]:
```suggestion
        VectorMask<$Boxtype$> blendMask = iota.toVector().compare(VectorOperators.LT, (broadcast(($type$)(length() - origin))));
```

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 7]
```diff
@@ -2279,6 +2276,17 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
     public abstract
     $abstractvectortype$ slice(int origin);
 
+    /*package-private*/
+    final
+    @ForceInline
+    $abstractvectortype$ sliceTemplate(int origin) {
+        Objects.checkIndex(origin, length() + 1);
+        VectorShuffle<$Boxtype$> iota = iotaShuffle();
+        VectorMask<$Boxtype$> BlendMask = iota.toVector().compare(VectorOperators.LT, (broadcast(($type$)(length() - origin))));
```
Comment (PaulSandoz) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2285]:
```suggestion
        VectorMask<$Boxtype$> blendMask = iota.toVector().compare(VectorOperators.LT, (broadcast(($type$)(length() - origin))));
```

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 8]
```diff
@@ -2293,21 +2301,12 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
...
-        case 1:
-            System.arraycopy(slice, firstPart, res, 0, origin);
-            break;
-        default:
-            throw wrongPartForSlice(part);
-        }
-        return vectorFactory(res);
+        Objects.checkIndex(origin, length() + 1);
+        VectorShuffle<$Boxtype$> iota = iotaShuffle();
+        VectorMask<$Boxtype$> BlendMask = iota.toVector().compare((part == 0) ? VectorOperators.GE : VectorOperators.LT,
```
Comment (PaulSandoz) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2306]:
```suggestion
        VectorMask<$Boxtype$> blendMask = iota.toVector().compare((part == 0) ? VectorOperators.GE : VectorOperators.LT,
```

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 9]
```diff
@@ -2293,21 +2301,12 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
...
+                                                                  (broadcast(($type$)(origin))));
+        iota = iotaShuffle(-origin, 1, true);
+        return (($abstractvectortype$)w).blend(this.rearrange(iota), BlendMask);
```
Comment (PaulSandoz) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2309]:
```suggestion
        return that.blend(this.rearrange(iota), blendMask);
```

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 10]
```diff
@@ -2279,6 +2276,17 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
...
 
+    /*package-private*/
+    final
+    @ForceInline
+    $abstractvectortype$ sliceTemplate(int origin) {
+        Objects.checkIndex(origin, length() + 1);
+        VectorShuffle<$Boxtype$> iota = iotaShuffle();
+        VectorMask<$Boxtype$> BlendMask = iota.toVector().compare(VectorOperators.LT, (broadcast(($type$)(length() - origin))));
+        iota = iotaShuffle(origin, 1, true);
+        return vspecies().zero().blend(this.rearrange(iota), BlendMask);
```
Comment (PaulSandoz) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2287]:
```suggestion
        return vspecies().zero().blend(this.rearrange(iota), blendMask);
```

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 11]
```diff
@@ -2337,6 +2336,19 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
...
     $abstractvectortype$ unslice(int origin);
 
+    /*package-private*/
+    final
+    @ForceInline
+    $abstractvectortype$
+    unsliceTemplate(int origin) {
+        Objects.checkIndex(origin, length() + 1);
+        VectorShuffle<$Boxtype$> iota = iotaShuffle();
+        VectorMask<$Boxtype$> BlendMask = iota.toVector().compare(VectorOperators.GE,
```
Comment (PaulSandoz) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2346]:
```suggestion
        VectorMask<$Boxtype$> blendMask = iota.toVector().compare(VectorOperators.GE,
```

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template [hunk 12]
```diff
@@ -2337,6 +2336,19 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
...
+                                                                  (broadcast(($type$)(origin))));
+        iota = iotaShuffle(-origin, 1, true);
+        return vspecies().zero().blend(this.rearrange(iota), BlendMask);
```
Comment (PaulSandoz) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:2349]:
```suggestion
        return vspecies().zero().blend(this.rearrange(iota), blendMask);
```

**Comment by sviswa7** [comment]
@PaulSandoz Thanks a lot for these suggestions. I have included all of them in this updated commit.

**Review by PaulSandoz — APPROVED** [review]

**Comment by sviswa7** [comment]
@PaulSandoz would it be possible for you to run this through your testing?

**Comment by PaulSandoz** [comment]
> @PaulSandoz would it be possible for you to run this through your testing?

Started, will report back when done.

**Comment by PaulSandoz** [comment]
> > @PaulSandoz would it be possible for you to run this through your testing?
> 
> Started, will report back when done.

Tier 1 to 3 tests pass all pass on build profiles linux-x64 linux-aarch64 macosx-x64 windows-x64

**Comment by sviswa7** [comment]
> > > @PaulSandoz would it be possible for you to run this through your testing?
> > 
> > 
> > Started, will report back when done.
> 
> Tier 1 to 3 tests all pass on build profiles linux-x64 linux-aarch64 macosx-x64 windows-x64

@PaulSandoz Thanks a lot!

**Comment by sviswa7** [comment]
@iwanowww Could you please review if the change in library_call.cpp looks ok to you?

**Comment by sviswa7** [comment]
@PaulSandoz After we fixed the Objects.checkIndex arguments per your review comment, the changes in library_call.cpp are no more needed so I have backed them out.  That leaves only changes in vector api code which you have reviewed. Please let me know if it is ok to push.

**Review by iwanowww — APPROVED** [review]
Looks good.

PS: I still think there's a problem with `LibraryCallKit::inline_preconditions_checkIndex`: it shouldn't bail out intrinsification in the middle of the process leaving a partially constructed graph behind. I don't see why short-circuiting the logic once the path is dead (`if (stopped()) return true;`) won't work. But that's a topic for another fix.

**Comment by sviswa7** [comment]
Thanks a lot for the review Paul and Vladimir.
Vladimir, I will submit a separate patch for LibraryCallKit::inline_preconditions_checkIndex fix.

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java [hunk 13]
```diff
@@ -2024,21 +2032,12 @@ ByteVector slice(int origin,
...
-        int firstPart = vlen - origin;
-        switch (part) {
-        case 0:
-            System.arraycopy(slice, 0, res, origin, firstPart);
-            break;
-        case 1:
-            System.arraycopy(slice, firstPart, res, 0, origin);
-            break;
-        default:
-            throw wrongPartForSlice(part);
```
Comment (eme64) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java:2039]: Did we "optimize away" the out of bounds check in this patch?
I think after this patch `wrongPartForSlice` is dead code.
I'm going to fix it with https://github.com/openjdk/jdk/pull/30113, and add some tests too.

Comment (sviswa7) [reply src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java:2039]: There is a call to Objects.checkIndex() in sliceTemplate() which does out of bounds check (e.g. ByteVector.java).

Comment (sviswa7) [reply src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java:2039]: I think you are referring to part number, thanks for taking care of that.
